### PR TITLE
Stabilize todo macro

### PR DESCRIPTION
**Pull request by ghost**
The `todo!` macro is just another name for `unimplemented!`.

Tracking issue: https://github.com/rust-lang/rust/issues/59277

This PR needs a FCP to merge.

r? @withoutboats

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by withoutboats** [comment]
@rfcbot fcp merge

Like Stjepan, I've been using `todo`, which is so much more pleasant than `unimplemented` (and even than `panic`). I like having this alias in std, and I want to be able to use it without turning on a feature flag.

**Comment by alexcrichton** [comment]
@rfcbot concern plan-for-unimplemented

Is there a plan for what to do with `unimplemented!`? Is the intended long-term effect that we have both macros, or is the intended long-term goal that we deprecate `unimplemented!` in favor of `todo!`?

**Comment by stjepang** [comment]
@alexcrichton I'm quoting @withoutboats here:

https://github.com/rust-lang/rust/pull/56348#issuecomment-466169688
https://github.com/rust-lang/rust/pull/56348#issuecomment-473922040

The short answer is that we keep `unimplemented!` and don't really make a distinction from `todo!`. Which of these two macros you use is a matter of personal taste.

**Comment by withoutboats** [comment]
> Is there a plan for what to do with unimplemented!? Is the intended long-term effect that we have both macros, or is the intended long-term goal that we deprecate unimplemented! in favor of todo!?

I don't think its worthwhile to deprecate unimplemented. Many people already use just `panic` for this purpose and this is totally fine. We really don't need to dictate to people how they write their "panic because i didnt write this code yet" code.

**Comment by alexcrichton** [comment]
Ok, thanks for the clarification! I'm not a huge fan of having multiple ways to do things like this, but this is basically just a convenience and conversions also are a matter of taste with libstd, so it seems reasonable enough to me.

@rfcbot resolve plan-for-unimplemented

**Comment by JohnBSmith** [comment]
I think it makes sense that the two macros could have slightly different meanings, in that `todo!()` would indicate work in progress, but `unimplemented!()` would indicate something omitted intentionally.

For example, if in a textbook some detail of an algorithm does not matter for the current consideration, an `unimplemented!()` could be placed rather than `todo!()`.

**Comment by smmalis37** [comment]
Similar to @alexcrichton I'm not a fan of multiple ways to do something simple like this. Is there really a need for this over something like `unimplemented!("TODO: Write this.")` vs `unimplemented!("Omitted for this example.")`?

I feel like having both of these makes it harder for people to track all instances of these macros for not much gain in expressiveness as they now have to know that there's two possibilities, not just one.

**Comment by stjepang** [comment]
@smmalis37

Well, I like prototyping Rust code by defining functions without implementing their bodies, leaving `unimplemented!()` in their place. But the thing is, if I'm writing code for myself, I tend to use `panic!()` instead because it's just so much easier to type than the other macro. In fact, it feels almost as if Rust is discouraging me from prototyping code.

Yes, this can be solved by external crates but I'm not going to add a dependency just for a single shorter macro. And yes, it might be solved by text editor's help but that isn't a solution everywhere (think Rust Playground).

So the bottom line is that I'm a human and having the `todo!()` macro would make me a happier Rust programmer. I feel the drawback of having multiple ways of doing the same thing is a small price to pay for that, but also understand others feel differently. *shrug*

**Comment by petrochenkov** [comment]
Duplication like this makes people learning the library go through the next process
- Should I use `unimplemented` or `todo`?
- Google -> reading top relevant Stack Overflow question "`unimplemented` vs `todo`"
- Top answer: they are exactly the same
- Questions in comments - "but why???", another answer contains detailed archeology with links to this thread and other thread.
- (Some percent of especially enthusiastic users goes back to the issue tracker or user forum to report that the library is a mess.)

Is this ok? I'm not sure, probably not a big deal, but I can't say I like it too much.

**Comment by petrochenkov** [comment]
On a related note, we really need lints with arguments enabling targeted deprecation (like `#![allow(deprecated(unimplemented))]`) to deprecate things like `unimplemented!()`.

**Comment by matklad** [comment]
>On a related note, we really need lints with arguments enabling targeted deprecation

While we are at it, Kotlin's [`replaceWith`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-replace-with/index.html) and associated tooling is also a must-have.

https://hackernoon.com/how-kotlins-deprecated-relieves-pain-of-colossal-refactoring-8577545aaed

**Comment by withoutboats** [comment]
I really can't relate to these comments when std already has several synonyms for panic. I think @alexcrichton's opinion might be that none of these should have been added (since I remember him saying once that you should just use panic instead of unimplemented or unreachable), but we're well past that point.

**Comment by totorigolo** [comment]
I would just like to add that some of my `TODO` comments are more `debug!("TODO this")` than `panic!("TODO this")`. So now I will have to choose if I want to use a panic-ing macro or a noop-ing comment, depending on the context. Both have the same name, which is kinda confusing.

Fortunately, I can override it to what I want... except that I shouldn't because then it'll be hard to know it's behaviour when switching between crates.

**Comment by alexcrichton** [comment]
To clarify my opinion, I'm not a huge fan of multiple ways to do the same thing and I think it's probably safe to say almost everyone is in roughly the same boat. That being religiously striving after this goal can and has been harmful in the past. 

The standard library in general does a great job of not providing lots of tools for the same option but an area this is definitely not true is conversions. For example you can count the number of ways to convert `&String` to `&str` and it's quite large! This isn't really a bad thing though in my opinion.

I think it's of course the case that @petrochenkov's example of doom and gloom can come up, but no matter what we do that comes up. People will be mildly confused about these two macros, yes, but the whole point of them is convenience. Having slightly different conveniences isn't the end of the world and if folks are motivated enough to push it through then that seems fine by me at least.

**Comment by Amanieu** [comment]
I use `unimplemented!` quite liberally when I am developing code. With auto-completion in vscode only 3 keystrokes are needed (`un<TAB`), so I personally don't see a pressing need for a new macro which is the same as the existing one except shorter to type. Maybe the situation is different for people using other editors?

**Comment by mark-i-m** [comment]
I use (neo)vim + rust-analyzer, which ATM doesn't tend to complete macro names, and the shorter macro would definitely make my life easier. Also, I already grep for TODO, when working, so having the macro with that name would be a nice bonus...

**Comment by VictorGavrish** [comment]
What are the actual benefits of adding this alias? I don't see it clearly enumerated anywhere.

I've skimmed through some of the previous discussion, and I think most of the arguments I've seen in favor of this change have been of the "why not?" variety. And indeed, I don't see a big drawback to doing it. In fact, it's mostly "meh" from me. But I'd argue that any addition to the standard library should have a firm "why yes" argument for it, rather than an absence of a strong resistance. There should be a certain bar for any new thing. As a casual Rust user, I've got the impression that this bar exists, and it's pretty high. But doing this change would seem inconsistent with that.

Is it just that `todo!()` is easier to type than `unimplemented!()`? That doesn't seem to be a good enough reason to add anything to the standard library. Also, I'd argue that the difference in typing (even without completion assistance) is dwarfed by the unwieldy `!()`. That thing takes me twice as long to type as the word itself because that's two different `shift` keys that I need to press and hold. Is it just that the name fits better? I think that's not very apparent and might be a matter of taste. Do we want to add different ways to do one thing just to suit people who have different tastes? Again, that doesn't seem to be like a strong enough reason to intentionally add duplication.

Just my two cents.

**Comment by mark-i-m** [comment]
For me the benefits are

1. easier to type (though I agree that !() is a pain to type). It might be just me, but one rust's great unsung features is brevity IMHO.
2. I already grep for TODO, so this streamlines my workflow signficantly.
3. unambiguous meaning: its clear that a todo!() is something I intend to do if others come across it, whereas panic and unimplemented do not communicate that idea.

**Comment by anowlcalledjosh** [comment]
fwiw, I suspect I just type weirdly, but `todo!()` and particularly `!()` are very _easy_ for me to type – `todo` is a quick alternation between left and right hands, and `!()` is just left-shift plus `!` (left hand) and `()` (right hand). In contrast, the first six letters of `unimplemented` are all typed with the right hand, which is comparatively slow and awkward.

File: src/libcore/macros.rs [hunk 1]
```diff
@@ -614,7 +618,7 @@ macro_rules! unimplemented {
 /// }
 /// ```
 #[macro_export]
-#[unstable(feature = "todo_macro", issue = "59277")]
+#[stable(feature = "todo_macro", since = "1.37.0")]
```
Comment (Centril) [inline src/libcore/macros.rs]:
```suggestion
#[stable(feature = "todo_macro", since = "1.38.0")]
```

**Comment by Centril** [comment]
Since FCP has completed with a disposition to merge...

r? @Centril

r=me rollup with the change to the since version per suggestion above.

**Comment by alexcrichton** [comment]
This was contentious enough during FCP the libs team does not feel this is ready to merge, please do not r+. I will follow-up tomorrow when i have more time.

**Comment by alexcrichton** [comment]
@withoutboats or @stjepang as the proponents for stabilization here would y'all be ok responding to some of the more recent comments to address the concerns raised?

**Comment by Centril** [comment]
My personal take here is that this is a lot like `dbg!(...)`. While `dbg!(...)` does not have to exist in libstd, it is convenient and as a result it has been a resounding success that I see everyone use everywhere these days. It has even been an inspiration for development in other programming languages.

Both are macros clearly meant for temporary usage as a development aid. I think in that light convenience is what is key. While `todo!`'s marginal utility compared to `dbg!` isn't as high, it is still still adds to convenience by: a) being shorter (yes, this is important in this case!), b) using a familiar word one generally associates with stuff that needs to be done, c) easier to grep as @mark-i-m noted.

Personally, *I think `todo!` does not go far enough!* (though it is a good step). Eventually, I would be quite interested in pursuing language level [improvements](https://internals.rust-lang.org/t/micro-pre-rfc-todo-macro/8925/14?u=centril) to this [including](https://www.youtube.com/watch?v=UkDSL0U9ndQ) typed [holes](https://www.youtube.com/watch?v=mOtKD7ml0NU).

**Comment by jesskfullwood** [comment]
I don't have any particular insights but ever since this was introduced I've been using it everywhere, much easier to type than `unimplemented` even though muscle-memory is starting to work for the latter.

I know the intention is that semantically they are identical, but I also think that `todo` more directly expresses my usual meaning. It is active - this is something I will do! `unimplemented` sounds passive by comparison.

**Comment by smmalis37** [comment]
Also to clarify my opinion: if the proposal was "Deprecate `uimplemented!` in favor of `todo!`" I'd be much more in favor of it (aside from the obvious backwards compatibility problems). However since `unimplemented!` is already stabilized I'm strongly against adding synonyms that make it harder to do a project-wide search for all instances. If `todo!` actually did something that `unimplemented!` can't or shouldn't then I'd have to reevaluate based on what that functionality was.

**Comment by withoutboats** [comment]
> @withoutboats or @stjepang as the proponents for stabilization here would y'all be ok responding to some of the more recent comments to address the concerns raised?

I don't really know what you want here, I feel both sides of the arguments have been made here and the winner is pretty clear. I'm completely confounded that the idea that we are adding synonyms and that's bad has any traction on the libs team, when are literally just adding the nth synonym for panic to the set of synonyms we already have - the downside already exists, I myself write `panic!("todo")` because it flows more naturally for me than `unimplemented!()`. I think the upsides have been well articulated previously and I would like this to just be stabilized already.

**Comment by alexcrichton** [comment]
I think at this point this PR just needs someone to drive it, and other members of the libs team (including myself) I believe do not want to drive this. (but I could be wrong!)

**Comment by Centril** [comment]
@alexcrichton What does driving entail here?

**Comment by alexcrichton** [comment]
I don't really have a concrete checklist of what it might entail but I'm pretty certain that if this is being driven then I wouldn't have to personally be involved so often. I would suspect that a driver would take the shape of what a shepherd looks like for the lang team, engaging those interested on this thread from both angles of merging and not merging and attempting to reach consensus. I cannot personally do this at this time.

**Comment by withoutboats** [comment]
The conversation has reached a steady state at which the team just has to make a decision; I don't believe there are arguments to be made. I've proposed to merge, boxes have been checked, FCP has been completed, but the rest of the team has said they're not comfortable merging this. I'm happy to r+ this if that's what you want me to do.

**Comment by withoutboats** [comment]
Basically I think someone on the libs team needs to champion *not merging this* by filing a blocking concern to move this out of FCP. If no one is willing to do this, as far as I'm concerned we've had the discussion  and followed the process to which this can now be merged.

**Comment by alexcrichton** [comment]
I am not personally comfortable r+'ing this, but I'm only one of many members on the libs team. I do not personally care what happens with this at this point.

**Comment by Centril** [comment]
(Just a note re. @rfcbot -- If I recall the bot's code correctly, I don't believe it will listen to `concern`, `cancel`, etc. if you do it in FCP-completed state nor will closing and reopening remove FCP-completed.)

**Comment by withoutboats** [comment]
I'll give other members of the team a few days to see if anyone wants to champion the other side, and if they do we can work together to try to reach a consensus. If next Monday this is still FCP-complete, I'll r+ it.

(@Centril thanks for the heads up, I think we can handle reading the thread for concerns ourselves on this one).

**Comment by withoutboats** [comment]
I'd said I'd r+ this in 5 days 15 days ago, and no one has offered to champion a concern against this. So:

@bors r+

Obviously many panic synonyms is not ideal, but we crossed that bridge with `unimplemented` and `unreachable`. The problem with those two synonyms is that while they can make code self-documenting, they're so much longer than `panic` that many users don't use them. `todo` being brief makes it much more handy for prototyping that `unimplemented`.

There's been some discussion around trying to define clearly when a user should use `unimplemented` instead of `todo`, or of deprecating `unimplemented`. However, these definitions usually seem to make `unimplemented` overlap with the use case for `unreachable`, and deprecating a prototyping tool that isn't supposed to survive long term in your code seems like churn without good motivation.

**Comment by SimonSapin** [comment]
In any future stabilization PR, please remember to [make it close](https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords) the corresponding tracking issue. Thanks!

**Pull request merged**